# 1.Data Preprocessing

전처리 준비를 위한 디렉토리 세팅
- 현재 실행중인 ipynb 파일과 zip 폴더가 같은 경로에 있으면 됨
- FOLDER(폴더명 상관 없음)
    - .git
    - FILENAME.ipynb(현재 파일)
    - zip/
        - Abuse/
        - Arrest/
        - Arson/
        - Assault/
        - Burglary/
        - Explosion/
        - Fighting/
        - Normal_Videos-Part-1/
    - (NP)

## 1.1 Transform Video into Numpy Array
    mp4file -> Numpy
    save as 
     './NP/Abuse',
     './NP/Normal-Videos-Part-1',
     './NP/Assault',
     './NP/Arson',
     './NP/Burglary',
     './NP/Fighting',
     './NP/Explosion',
     './NP/Arrest

In [1]:
import cv2  

import numpy as np
import matplotlib.pyplot as plt

import os 
from tqdm import tqdm

In [2]:
def Video2Npy(file_path, resize=(128,128)): # 비디오파일에서 프레임 개수만큼 이미지 추출 후 Npy로 변경

    cap = cv2.VideoCapture(file_path) # 해당 경로 파일로부터 VideoCapture 객체 생성
    
    len_frames = int(cap.get(7)) # 총 프레임 수
    
    # 프레임 추출
    try:
        frames = []
        for i in range(len_frames - 1):
            _, frame = cap.read() # 비디오를 한 프레임씩 읽음

            frame = cv2.resize(frame, resize, interpolation=cv2.INTER_AREA) # 사이즈 다운 시 보간법 사용 : cv2.INTER_AREA
            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB) # RGB
            frame = np.reshape(frame, (128,128,3)) # RGB 3채널 벡터로 변환
            frames.append(frame)
    except:
        print("Error: ", file_path, len_frames, i)
    finally:
        frames = np.array(frames)
        cap.release() # VideoCapture 객체 해제

    
    return frames

In [3]:
def Save2Npy(file_dir, save_dir): # Video2Npy 결과(.npy)를 디렉토리에 저장

  
    if not os.path.exists(save_dir): # 저장 디렉토리 생성
        os.makedirs(save_dir)

    videos = os.listdir(file_dir) 

    for v in tqdm(videos):
        # Split video name
        video_name = v.split('.')[0]

        # Get src 
        video_path = os.path.join(file_dir, v)

        # Get dest 
        save_path = os.path.join(save_dir, video_name + '.npy') 
        print(video_name)

        # Load and preprocess video
        data = Video2Npy(file_path=video_path, resize=(128,128))
        data = np.uint8(data)
        
        # Save as .npy file
        np.save(save_path, data)
    
    return None

In [6]:
import os

dirlist = []

for dirname, _, filenames in os.walk('./zip'):
    dirlist.append(os.path.join(dirname))

dirlist

# os.walk
# os.walk() 인자로 전달된 path에 대해서 다음 3개의 값이 있는 tuple을 넘겨줍니다.
# root : dir와 files가 있는 path
# dirs : root 아래에 있는 폴더들
# files : root 아래에 있는 파일들

['./zip',
 './zip/Abuse',
 './zip/Assault',
 './zip/Arson',
 './zip/Burglary',
 './zip/Normal_Videos-Part-1',
 './zip/Fighting',
 './zip/Explosion',
 './zip/Arrest']

In [8]:
if not os.path.exists('./NP'):   
    for i in range(1, 9): # 
        source_path = dirlist[i]
        target_path = './NP/' + dirlist[i].split('/')[2] 

        Save2Npy(file_dir=source_path, save_dir=target_path)
else:
    print('Already done')

  0%|          | 0/50 [00:00<?, ?it/s]

Abuse048_x264


  2%|▏         | 1/50 [00:03<02:42,  3.33s/it]

Abuse041_x264


  4%|▍         | 2/50 [00:04<01:38,  2.04s/it]

Abuse035_x264


  6%|▌         | 3/50 [00:06<01:39,  2.12s/it]

Abuse032_x264


  8%|▊         | 4/50 [00:10<02:11,  2.85s/it]

Abuse027_x264


 10%|█         | 5/50 [00:14<02:22,  3.16s/it]

Abuse042_x264


 12%|█▏        | 6/50 [00:40<07:58, 10.87s/it]

Abuse044_x264


 14%|█▍        | 7/50 [00:43<05:59,  8.36s/it]

Abuse033_x264


 16%|█▌        | 8/50 [00:45<04:29,  6.43s/it]

Abuse007_x264


 18%|█▊        | 9/50 [00:46<03:13,  4.71s/it]

Abuse039_x264


 20%|██        | 10/50 [00:59<04:45,  7.13s/it]

Abuse019_x264


 22%|██▏       | 11/50 [01:02<03:50,  5.92s/it]

Abuse029_x264


 24%|██▍       | 12/50 [01:02<02:39,  4.20s/it]

Abuse012_x264


 26%|██▌       | 13/50 [01:06<02:34,  4.18s/it]

Abuse028_x264


 28%|██▊       | 14/50 [01:07<01:57,  3.26s/it]

Abuse013_x264


 30%|███       | 15/50 [01:11<02:01,  3.47s/it]

Abuse008_x264


 32%|███▏      | 16/50 [01:18<02:31,  4.46s/it]

Abuse021_x264


 34%|███▍      | 17/50 [01:20<01:58,  3.59s/it]

Abuse004_x264


 36%|███▌      | 18/50 [01:33<03:31,  6.61s/it]

Abuse009_x264


 38%|███▊      | 19/50 [01:34<02:31,  4.87s/it]

Abuse030_x264


 40%|████      | 20/50 [01:35<01:54,  3.80s/it]

Abuse024_x264


 42%|████▏     | 21/50 [01:36<01:24,  2.91s/it]

Abuse036_x264


 44%|████▍     | 22/50 [01:40<01:29,  3.20s/it]

Abuse006_x264


 46%|████▌     | 23/50 [01:44<01:29,  3.33s/it]

Abuse010_x264


 48%|████▊     | 24/50 [01:45<01:09,  2.69s/it]

Abuse005_x264


 50%|█████     | 25/50 [01:46<00:55,  2.23s/it]

Abuse017_x264


 52%|█████▏    | 26/50 [01:48<00:53,  2.23s/it]

Abuse049_x264


 54%|█████▍    | 27/50 [01:51<00:53,  2.30s/it]

Abuse011_x264


 56%|█████▌    | 28/50 [01:54<00:56,  2.56s/it]

Abuse040_x264


 58%|█████▊    | 29/50 [01:57<00:54,  2.62s/it]

Abuse034_x264


 60%|██████    | 30/50 [01:58<00:44,  2.21s/it]

Abuse047_x264


 62%|██████▏   | 31/50 [01:59<00:36,  1.93s/it]

Abuse020_x264


 64%|██████▍   | 32/50 [02:02<00:37,  2.10s/it]

Abuse037_x264


 66%|██████▌   | 33/50 [02:03<00:32,  1.90s/it]

Abuse038_x264


 68%|██████▊   | 34/50 [02:04<00:24,  1.55s/it]

Abuse045_x264


 70%|███████   | 35/50 [02:08<00:35,  2.35s/it]

Abuse018_x264


 72%|███████▏  | 36/50 [02:10<00:31,  2.23s/it]

Abuse014_x264


 74%|███████▍  | 37/50 [02:12<00:28,  2.19s/it]

Abuse002_x264


 76%|███████▌  | 38/50 [02:13<00:20,  1.75s/it]

Abuse003_x264


 78%|███████▊  | 39/50 [02:16<00:23,  2.12s/it]

Abuse001_x264


 80%|████████  | 40/50 [02:18<00:21,  2.15s/it]

Abuse023_x264


 82%|████████▏ | 41/50 [02:19<00:15,  1.76s/it]

Abuse026_x264


 84%|████████▍ | 42/50 [02:21<00:13,  1.72s/it]

Abuse031_x264


 86%|████████▌ | 43/50 [02:24<00:15,  2.28s/it]

Abuse022_x264


 88%|████████▊ | 44/50 [02:27<00:13,  2.29s/it]

Abuse025_x264


 90%|█████████ | 45/50 [02:28<00:10,  2.10s/it]

Abuse015_x264


 92%|█████████▏| 46/50 [02:34<00:13,  3.35s/it]

Abuse046_x264


 94%|█████████▍| 47/50 [02:35<00:07,  2.63s/it]

Abuse016_x264


 96%|█████████▌| 48/50 [02:36<00:04,  2.18s/it]

Abuse043_x264


 98%|█████████▊| 49/50 [02:40<00:02,  2.46s/it]

Abuse050_x264


  0%|          | 0/50 [00:00<?, ?it/s]

Assault029_x264


  2%|▏         | 1/50 [00:00<00:24,  2.01it/s]

Assault051_x264


  4%|▍         | 2/50 [00:00<00:23,  2.03it/s]

Assault050_x264


  8%|▊         | 4/50 [00:01<00:16,  2.83it/s]

Assault038_x264
Assault004_x264


 10%|█         | 5/50 [00:04<00:53,  1.18s/it]

Assault033_x264


 12%|█▏        | 6/50 [00:05<00:49,  1.13s/it]

Assault040_x264


 14%|█▍        | 7/50 [00:08<01:15,  1.76s/it]

Assault007_x264


 16%|█▌        | 8/50 [00:09<01:11,  1.70s/it]

Assault006_x264


 18%|█▊        | 9/50 [00:16<02:11,  3.20s/it]

Assault001_x264


 20%|██        | 10/50 [00:18<01:53,  2.83s/it]

Assault003_x264


 22%|██▏       | 11/50 [00:22<01:59,  3.07s/it]

Assault044_x264


 24%|██▍       | 12/50 [00:22<01:28,  2.33s/it]

Assault024_x264


 26%|██▌       | 13/50 [00:23<01:04,  1.73s/it]

Assault048_x264


 28%|██▊       | 14/50 [00:24<00:59,  1.65s/it]

Assault025_x264


 30%|███       | 15/50 [00:25<00:50,  1.44s/it]

Assault026_x264


 32%|███▏      | 16/50 [00:26<00:41,  1.21s/it]

Assault005_x264


 34%|███▍      | 17/50 [00:27<00:37,  1.14s/it]

Assault011_x264


 36%|███▌      | 18/50 [00:29<00:43,  1.36s/it]

Assault027_x264


 38%|███▊      | 19/50 [00:34<01:21,  2.63s/it]

Assault032_x264


 40%|████      | 20/50 [00:35<00:59,  1.98s/it]

Assault037_x264


 42%|████▏     | 21/50 [00:37<01:03,  2.18s/it]

Assault018_x264


 44%|████▍     | 22/50 [00:38<00:45,  1.62s/it]

Assault002_x264


 46%|████▌     | 23/50 [00:40<00:47,  1.75s/it]

Assault034_x264


 48%|████▊     | 24/50 [00:41<00:46,  1.79s/it]

Assault039_x264


 50%|█████     | 25/50 [00:42<00:33,  1.32s/it]

Assault015_x264


 52%|█████▏    | 26/50 [00:43<00:30,  1.29s/it]

Assault023_x264


 54%|█████▍    | 27/50 [00:44<00:26,  1.15s/it]

Assault012_x264


 56%|█████▌    | 28/50 [00:44<00:22,  1.00s/it]

Assault049_x264


 58%|█████▊    | 29/50 [00:45<00:21,  1.03s/it]

Assault017_x264


 60%|██████    | 30/50 [00:48<00:31,  1.57s/it]

Assault052_x264


 62%|██████▏   | 31/50 [00:49<00:25,  1.32s/it]

Assault035_x264


 64%|██████▍   | 32/50 [00:50<00:21,  1.19s/it]

Assault030_x264


 66%|██████▌   | 33/50 [00:52<00:24,  1.42s/it]

Assault031_x264


 68%|██████▊   | 34/50 [00:53<00:22,  1.40s/it]

Assault045_x264


 70%|███████   | 35/50 [00:56<00:25,  1.69s/it]

Assault020_x264


 72%|███████▏  | 36/50 [00:57<00:20,  1.49s/it]

Assault019_x264


 74%|███████▍  | 37/50 [01:02<00:32,  2.52s/it]

Assault046_x264


 76%|███████▌  | 38/50 [01:02<00:24,  2.01s/it]

Assault028_x264


 78%|███████▊  | 39/50 [01:05<00:23,  2.10s/it]

Assault016_x264


 80%|████████  | 40/50 [01:11<00:34,  3.45s/it]

Assault047_x264


 82%|████████▏ | 41/50 [01:14<00:29,  3.29s/it]

Assault036_x264


 84%|████████▍ | 42/50 [01:15<00:20,  2.53s/it]

Assault008_x264


 86%|████████▌ | 43/50 [01:21<00:25,  3.65s/it]

Assault013_x264


 88%|████████▊ | 44/50 [01:24<00:19,  3.27s/it]

Assault010_x264


 90%|█████████ | 45/50 [01:38<00:33,  6.66s/it]

Assault042_x264


 92%|█████████▏| 46/50 [01:47<00:29,  7.39s/it]

Assault009_x264


 94%|█████████▍| 47/50 [01:50<00:18,  6.06s/it]

Assault041_x264


 96%|█████████▌| 48/50 [01:54<00:10,  5.47s/it]

Assault022_x264


 98%|█████████▊| 49/50 [01:56<00:04,  4.44s/it]

Assault014_x264


  0%|          | 0/50 [00:00<?, ?it/s]

Arson009_x264


  2%|▏         | 1/50 [00:00<00:30,  1.62it/s]

Arson022_x264


  4%|▍         | 2/50 [00:07<03:39,  4.57s/it]

Arson025_x264


  6%|▌         | 3/50 [00:13<03:46,  4.82s/it]

Arson013_x264


  8%|▊         | 4/50 [00:18<03:48,  4.98s/it]

Arson049_x264


 10%|█         | 5/50 [00:18<02:28,  3.30s/it]

Arson011_x264


 12%|█▏        | 6/50 [00:19<01:51,  2.53s/it]

Arson038_x264


 14%|█▍        | 7/50 [00:20<01:21,  1.90s/it]

Arson050_x264


 16%|█▌        | 8/50 [00:21<01:12,  1.72s/it]

Arson006_x264


 18%|█▊        | 9/50 [00:23<01:17,  1.90s/it]

Arson029_x264


 20%|██        | 10/50 [00:24<01:06,  1.65s/it]

Arson041_x264


 22%|██▏       | 11/50 [00:28<01:21,  2.09s/it]

Arson007_x264


 24%|██▍       | 12/50 [00:33<01:54,  3.01s/it]

Arson045_x264


 26%|██▌       | 13/50 [00:36<01:53,  3.06s/it]

Arson023_x264


 28%|██▊       | 14/50 [00:37<01:29,  2.47s/it]

Arson044_x264


 30%|███       | 15/50 [00:38<01:10,  2.02s/it]

Arson015_x264


 32%|███▏      | 16/50 [00:46<02:14,  3.97s/it]

Arson005_x264


 34%|███▍      | 17/50 [00:47<01:41,  3.07s/it]

Arson031_x264


 36%|███▌      | 18/50 [00:50<01:31,  2.86s/it]

Arson014_x264


 38%|███▊      | 19/50 [00:52<01:22,  2.65s/it]

Arson021_x264


 40%|████      | 20/50 [00:53<01:06,  2.21s/it]

Arson051_x264


 42%|████▏     | 21/50 [00:54<00:52,  1.81s/it]

Arson019_x264


 44%|████▍     | 22/50 [03:04<18:44, 40.17s/it]

Arson046_x264


 46%|████▌     | 23/50 [03:08<13:14, 29.42s/it]

Arson037_x264


 48%|████▊     | 24/50 [03:10<09:09, 21.13s/it]

Arson030_x264


 50%|█████     | 25/50 [03:18<07:13, 17.35s/it]

Arson048_x264


 52%|█████▏    | 26/50 [03:19<04:58, 12.42s/it]

Arson042_x264


 54%|█████▍    | 27/50 [03:30<04:35, 11.97s/it]

Arson028_x264


 56%|█████▌    | 28/50 [03:35<03:35,  9.80s/it]

Arson039_x264


 58%|█████▊    | 29/50 [03:41<03:05,  8.82s/it]

Arson001_x264


 60%|██████    | 30/50 [03:50<02:53,  8.67s/it]

Arson003_x264


 62%|██████▏   | 31/50 [03:57<02:38,  8.34s/it]

Arson052_x264


 64%|██████▍   | 32/50 [04:08<02:41,  8.97s/it]

Arson035_x264


 66%|██████▌   | 33/50 [04:10<02:00,  7.10s/it]

Arson027_x264


 68%|██████▊   | 34/50 [04:16<01:44,  6.55s/it]

Arson036_x264


 70%|███████   | 35/50 [04:19<01:23,  5.54s/it]

Arson034_x264


 72%|███████▏  | 36/50 [04:23<01:12,  5.19s/it]

Arson047_x264


 74%|███████▍  | 37/50 [04:27<00:59,  4.60s/it]

Arson016_x264


 76%|███████▌  | 38/50 [04:30<00:51,  4.32s/it]

Arson010_x264


 78%|███████▊  | 39/50 [04:36<00:53,  4.90s/it]

Arson053_x264


 80%|████████  | 40/50 [04:43<00:54,  5.47s/it]

Arson017_x264


 82%|████████▏ | 41/50 [04:44<00:37,  4.18s/it]

Arson008_x264


 84%|████████▍ | 42/50 [04:46<00:26,  3.37s/it]

Arson020_x264


 86%|████████▌ | 43/50 [04:48<00:21,  3.08s/it]

Arson024_x264


 88%|████████▊ | 44/50 [04:51<00:17,  2.99s/it]

Arson012_x264


 90%|█████████ | 45/50 [04:52<00:11,  2.31s/it]

Arson002_x264


 92%|█████████▏| 46/50 [04:56<00:10,  2.73s/it]

Arson026_x264


 94%|█████████▍| 47/50 [05:00<00:10,  3.36s/it]

Arson018_x264


 96%|█████████▌| 48/50 [05:01<00:05,  2.57s/it]

Arson040_x264


 98%|█████████▊| 49/50 [05:04<00:02,  2.61s/it]

Arson032_x264


  0%|          | 0/100 [00:00<?, ?it/s]

Burglary008_x264


  1%|          | 1/100 [00:04<07:10,  4.34s/it]

Burglary094_x264


  2%|▏         | 2/100 [00:05<04:09,  2.55s/it]

Burglary054_x264


  3%|▎         | 3/100 [00:06<02:32,  1.57s/it]

Burglary058_x264


  4%|▍         | 4/100 [00:17<08:38,  5.40s/it]

Burglary044_x264


  5%|▌         | 5/100 [00:23<08:49,  5.58s/it]

Burglary046_x264


  6%|▌         | 6/100 [00:25<06:46,  4.33s/it]

Burglary039_x264


  7%|▋         | 7/100 [00:33<08:32,  5.51s/it]

Burglary100_x264


  8%|▊         | 8/100 [00:39<08:41,  5.67s/it]

Burglary081_x264


  9%|▉         | 9/100 [00:39<06:04,  4.00s/it]

Burglary061_x264


 10%|█         | 10/100 [00:46<07:39,  5.11s/it]

Burglary032_x264


 11%|█         | 11/100 [01:01<12:04,  8.14s/it]

Burglary091_x264


 12%|█▏        | 12/100 [01:04<09:31,  6.50s/it]

Burglary099_x264


 13%|█▎        | 13/100 [01:05<07:02,  4.86s/it]

Burglary053_x264


 14%|█▍        | 14/100 [01:08<06:03,  4.23s/it]

Burglary037_x264


 15%|█▌        | 15/100 [01:10<04:51,  3.43s/it]

Burglary086_x264


 16%|█▌        | 16/100 [01:12<04:27,  3.19s/it]

Burglary093_x264


 17%|█▋        | 17/100 [01:13<03:20,  2.42s/it]

Burglary043_x264


 18%|█▊        | 18/100 [01:13<02:30,  1.84s/it]

Burglary068_x264


 19%|█▉        | 19/100 [01:14<01:58,  1.46s/it]

Burglary031_x264


 20%|██        | 20/100 [01:19<03:22,  2.54s/it]

Burglary051_x264


 21%|██        | 21/100 [01:20<02:45,  2.09s/it]

Burglary033_x264


 22%|██▏       | 22/100 [01:21<02:19,  1.79s/it]

Burglary014_x264


 23%|██▎       | 23/100 [01:23<02:27,  1.92s/it]

Burglary003_x264


 24%|██▍       | 24/100 [01:24<02:04,  1.64s/it]

Burglary038_x264


 25%|██▌       | 25/100 [01:25<01:43,  1.38s/it]

Burglary013_x264


 26%|██▌       | 26/100 [01:27<01:53,  1.54s/it]

Burglary036_x264


 27%|██▋       | 27/100 [01:28<01:37,  1.33s/it]

Burglary030_x264


 28%|██▊       | 28/100 [01:29<01:24,  1.17s/it]

Burglary002_x264


 29%|██▉       | 29/100 [01:31<01:51,  1.57s/it]

Burglary016_x264


 30%|███       | 30/100 [01:36<02:47,  2.39s/it]

Burglary023_x264


 31%|███       | 31/100 [01:46<05:25,  4.72s/it]

Burglary004_x264


 32%|███▏      | 32/100 [01:48<04:23,  3.88s/it]

Burglary074_x264


 33%|███▎      | 33/100 [01:49<03:38,  3.26s/it]

Burglary075_x264


 34%|███▍      | 34/100 [02:02<06:41,  6.08s/it]

Burglary065_x264


 35%|███▌      | 35/100 [02:10<07:04,  6.53s/it]

Burglary080_x264


 36%|███▌      | 36/100 [02:11<05:10,  4.86s/it]

Burglary034_x264


 37%|███▋      | 37/100 [02:12<03:53,  3.71s/it]

Burglary083_x264


 38%|███▊      | 38/100 [02:16<04:09,  4.03s/it]

Burglary041_x264


 39%|███▉      | 39/100 [02:18<03:20,  3.29s/it]

Burglary092_x264


 40%|████      | 40/100 [02:19<02:27,  2.46s/it]

Burglary057_x264


 41%|████      | 41/100 [02:23<03:01,  3.07s/it]

Burglary067_x264


 42%|████▏     | 42/100 [02:24<02:23,  2.48s/it]

Burglary042_x264


 43%|████▎     | 43/100 [02:28<02:37,  2.76s/it]

Burglary035_x264


 44%|████▍     | 44/100 [02:31<02:42,  2.91s/it]

Burglary084_x264


 45%|████▌     | 45/100 [02:34<02:52,  3.13s/it]

Burglary021_x264


 46%|████▌     | 46/100 [02:36<02:18,  2.57s/it]

Burglary077_x264


 47%|████▋     | 47/100 [02:37<02:00,  2.27s/it]

Burglary064_x264


 48%|████▊     | 48/100 [03:17<11:39, 13.44s/it]

Burglary048_x264


 49%|████▉     | 49/100 [03:18<08:23,  9.87s/it]

Burglary028_x264


 50%|█████     | 50/100 [03:19<05:57,  7.14s/it]

Burglary060_x264


 51%|█████     | 51/100 [03:20<04:14,  5.19s/it]

Burglary049_x264


 52%|█████▏    | 52/100 [03:28<04:47,  6.00s/it]

Burglary024_x264


 53%|█████▎    | 53/100 [03:31<04:06,  5.24s/it]

Burglary063_x264


 54%|█████▍    | 54/100 [03:38<04:26,  5.79s/it]

Burglary062_x264


 55%|█████▌    | 55/100 [03:51<05:57,  7.94s/it]

Burglary022_x264


 56%|█████▌    | 56/100 [03:54<04:38,  6.33s/it]

Burglary072_x264


 57%|█████▋    | 57/100 [03:56<03:38,  5.08s/it]

Burglary087_x264


 58%|█████▊    | 58/100 [03:58<02:53,  4.12s/it]

Burglary011_x264


 59%|█████▉    | 59/100 [03:59<02:18,  3.38s/it]

Burglary085_x264


 60%|██████    | 60/100 [04:02<02:03,  3.09s/it]

Burglary012_x264


 61%|██████    | 61/100 [04:03<01:40,  2.58s/it]

Burglary052_x264


 62%|██████▏   | 62/100 [04:12<02:49,  4.45s/it]

Burglary073_x264


 63%|██████▎   | 63/100 [04:14<02:20,  3.80s/it]

Burglary027_x264


 64%|██████▍   | 64/100 [04:17<02:03,  3.43s/it]

Burglary069_x264


 65%|██████▌   | 65/100 [04:19<01:42,  2.92s/it]

Burglary095_x264


 66%|██████▌   | 66/100 [04:55<07:20, 12.96s/it]

Burglary066_x264


 67%|██████▋   | 67/100 [04:57<05:19,  9.68s/it]

Burglary090_x264


 68%|██████▊   | 68/100 [05:00<04:05,  7.68s/it]

Burglary076_x264


 69%|██████▉   | 69/100 [05:11<04:27,  8.64s/it]

Burglary029_x264


 70%|███████   | 70/100 [05:15<03:39,  7.33s/it]

Burglary050_x264


 71%|███████   | 71/100 [05:16<02:37,  5.42s/it]

Burglary020_x264


 72%|███████▏  | 72/100 [05:19<02:09,  4.64s/it]

Burglary017_x264


 73%|███████▎  | 73/100 [05:21<01:41,  3.77s/it]

Burglary079_x264


 74%|███████▍  | 74/100 [05:33<02:46,  6.40s/it]

Burglary019_x264


 75%|███████▌  | 75/100 [05:36<02:10,  5.20s/it]

Burglary005_x264


 76%|███████▌  | 76/100 [05:42<02:14,  5.60s/it]

Burglary040_x264


 77%|███████▋  | 77/100 [05:47<02:02,  5.34s/it]

Burglary009_x264


 78%|███████▊  | 78/100 [05:49<01:37,  4.42s/it]

Burglary001_x264


 79%|███████▉  | 79/100 [05:54<01:34,  4.51s/it]

Burglary015_x264


 80%|████████  | 80/100 [05:55<01:10,  3.54s/it]

Burglary059_x264


 81%|████████  | 81/100 [05:56<00:50,  2.65s/it]

Burglary045_x264


 82%|████████▏ | 82/100 [06:05<01:24,  4.69s/it]

Burglary006_x264


 83%|████████▎ | 83/100 [06:13<01:36,  5.67s/it]

Burglary025_x264


 84%|████████▍ | 84/100 [06:18<01:28,  5.52s/it]

Burglary089_x264


 85%|████████▌ | 85/100 [06:21<01:11,  4.73s/it]

Burglary007_x264


 86%|████████▌ | 86/100 [06:22<00:50,  3.59s/it]

Burglary047_x264


 87%|████████▋ | 87/100 [06:25<00:42,  3.23s/it]

Burglary088_x264


 88%|████████▊ | 88/100 [06:29<00:41,  3.50s/it]

Burglary010_x264


 89%|████████▉ | 89/100 [06:34<00:43,  3.92s/it]

Burglary018_x264


 90%|█████████ | 90/100 [06:34<00:30,  3.02s/it]

Burglary096_x264


 91%|█████████ | 91/100 [06:40<00:34,  3.86s/it]

Burglary098_x264


 92%|█████████▏| 92/100 [06:41<00:23,  2.93s/it]

Burglary070_x264


 93%|█████████▎| 93/100 [06:42<00:15,  2.25s/it]

Burglary026_x264


 94%|█████████▍| 94/100 [06:43<00:11,  1.95s/it]

Burglary078_x264


 95%|█████████▌| 95/100 [06:43<00:07,  1.44s/it]

Burglary097_x264


 96%|█████████▌| 96/100 [06:43<00:04,  1.09s/it]

Burglary056_x264


 97%|█████████▋| 97/100 [06:45<00:03,  1.11s/it]

Burglary055_x264


 98%|█████████▊| 98/100 [06:47<00:02,  1.47s/it]

Burglary082_x264


 99%|█████████▉| 99/100 [06:49<00:01,  1.72s/it]

Burglary071_x264


  0%|          | 0/150 [00:00<?, ?it/s]

Normal_Videos_056_x264


  1%|          | 1/150 [00:01<03:11,  1.28s/it]

Normal_Videos_901_x264


  1%|▏         | 2/150 [00:02<02:41,  1.09s/it]

Normal_Videos_782_x264


  2%|▏         | 3/150 [00:06<06:36,  2.70s/it]

Normal_Videos_904_x264


  3%|▎         | 4/150 [00:07<04:41,  1.93s/it]

Normal_Videos_878_x264


  3%|▎         | 5/150 [00:07<03:10,  1.31s/it]

Normal_Videos_439_x264


  4%|▍         | 6/150 [00:11<04:57,  2.07s/it]

Normal_Videos_908_x264


  5%|▍         | 7/150 [00:12<03:53,  1.63s/it]

Normal_Videos_868_x264


  5%|▌         | 8/150 [00:14<04:07,  1.74s/it]

Normal_Videos_745_x264


  6%|▌         | 9/150 [00:14<03:00,  1.28s/it]

Normal_Videos_938_x264


  7%|▋         | 10/150 [00:18<04:57,  2.13s/it]

Normal_Videos_289_x264


  7%|▋         | 11/150 [00:19<03:55,  1.69s/it]

Normal_Videos_050_x264


  8%|▊         | 12/150 [00:22<05:08,  2.24s/it]

Normal_Videos_882_x264


  9%|▊         | 13/150 [00:23<04:30,  1.97s/it]

Normal_Videos_939_x264


  9%|▉         | 14/150 [00:24<03:34,  1.58s/it]

Normal_Videos_889_x264


 10%|█         | 15/150 [00:24<02:39,  1.18s/it]

Normal_Videos_634_x264


 11%|█         | 16/150 [00:36<09:35,  4.29s/it]

Normal_Videos_903_x264


 11%|█▏        | 17/150 [00:36<07:05,  3.20s/it]

Normal_Videos_401_x264


 12%|█▏        | 18/150 [00:38<05:48,  2.64s/it]

Normal_Videos_898_x264


 13%|█▎        | 19/150 [00:39<04:36,  2.11s/it]

Normal_Videos_907_x264


 13%|█▎        | 20/150 [00:39<03:32,  1.63s/it]

Normal_Videos_182_x264


 14%|█▍        | 21/150 [00:43<04:42,  2.19s/it]

Normal_Videos_875_x264


 15%|█▍        | 22/150 [00:45<04:37,  2.17s/it]

Normal_Videos_867_x264


 15%|█▌        | 23/150 [00:45<03:37,  1.71s/it]

Normal_Videos_934_x264


 16%|█▌        | 24/150 [00:47<03:36,  1.72s/it]

Normal_Videos_886_x264


 17%|█▋        | 25/150 [00:51<04:54,  2.36s/it]

Normal_Videos_006_x264


 17%|█▋        | 26/150 [00:52<04:01,  1.95s/it]

Normal_Videos_941_x264


 18%|█▊        | 27/150 [00:57<05:40,  2.77s/it]

Normal_Videos_884_x264


 19%|█▊        | 28/150 [01:06<09:49,  4.83s/it]

Normal_Videos_576_x264


 19%|█▉        | 29/150 [01:16<12:54,  6.40s/it]

Normal_Videos_014_x264


 20%|██        | 30/150 [01:18<09:43,  4.86s/it]

Normal_Videos_897_x264


 21%|██        | 31/150 [01:18<07:11,  3.63s/it]

Normal_Videos_780_x264


 21%|██▏       | 32/150 [01:20<05:58,  3.04s/it]

Normal_Videos_025_x264


 22%|██▏       | 33/150 [01:21<04:26,  2.28s/it]

Normal_Videos_129_x264


 23%|██▎       | 34/150 [01:21<03:18,  1.71s/it]

Normal_Videos_887_x264


 23%|██▎       | 35/150 [01:28<06:10,  3.23s/it]

Normal_Videos_801_x264


 24%|██▍       | 36/150 [01:30<05:36,  2.95s/it]

Normal_Videos_781_x264


 25%|██▍       | 37/150 [01:33<05:44,  3.04s/it]

Normal_Videos_778_x264


 25%|██▌       | 38/150 [01:34<04:34,  2.45s/it]

Normal_Videos_100_x264


 26%|██▌       | 39/150 [01:35<03:28,  1.88s/it]

Normal_Videos_912_x264


 27%|██▋       | 40/150 [01:36<02:46,  1.51s/it]

Normal_Videos_910_x264


 27%|██▋       | 41/150 [01:36<02:10,  1.20s/it]

Normal_Videos_914_x264


 28%|██▊       | 42/150 [01:37<01:54,  1.06s/it]

Normal_Videos_911_x264


 29%|██▊       | 43/150 [01:37<01:40,  1.07it/s]

Normal_Videos_717_x264


 29%|██▉       | 44/150 [01:39<01:44,  1.02it/s]

Normal_Videos_019_x264


 30%|███       | 45/150 [01:42<02:56,  1.68s/it]

Normal_Videos_872_x264


 31%|███       | 46/150 [01:43<02:25,  1.40s/it]

Normal_Videos_925_x264


 31%|███▏      | 47/150 [01:50<05:22,  3.14s/it]

Normal_Videos_893_x264


 32%|███▏      | 48/150 [01:55<06:37,  3.90s/it]

Normal_Videos_352_x264


 33%|███▎      | 49/150 [02:00<06:51,  4.08s/it]

Normal_Videos_247_x264


 33%|███▎      | 50/150 [02:07<08:23,  5.04s/it]

Normal_Videos_042_x264


 34%|███▍      | 51/150 [02:10<07:06,  4.31s/it]

Normal_Videos_895_x264


 35%|███▍      | 52/150 [02:12<06:08,  3.76s/it]

Normal_Videos_196_x264


 35%|███▌      | 53/150 [02:14<05:15,  3.25s/it]

Normal_Videos_931_x264


 36%|███▌      | 54/150 [02:16<04:32,  2.84s/it]

Normal_Videos_365_x264


 37%|███▋      | 55/150 [02:22<05:47,  3.66s/it]

Normal_Videos_900_x264


 37%|███▋      | 56/150 [02:23<04:34,  2.92s/it]

Normal_Videos_831_x264


 38%|███▊      | 57/150 [02:23<03:21,  2.16s/it]

Normal_Videos_936_x264


 39%|███▊      | 58/150 [02:24<02:45,  1.80s/it]

Normal_Videos_696_x264


 39%|███▉      | 59/150 [02:27<03:18,  2.18s/it]

Normal_Videos_453_x264


 40%|████      | 60/150 [02:32<04:23,  2.92s/it]

Normal_Videos_024_x264


 41%|████      | 61/150 [02:33<03:26,  2.32s/it]

Normal_Videos_758_x264


 41%|████▏     | 62/150 [02:34<02:57,  2.02s/it]

Normal_Videos_033_x264


 42%|████▏     | 63/150 [02:36<02:39,  1.83s/it]

Normal_Videos_015_x264


 43%|████▎     | 64/150 [02:36<02:00,  1.40s/it]

Normal_Videos_203_x264


 43%|████▎     | 65/150 [02:38<02:17,  1.62s/it]

Normal_Videos_725_x264


 44%|████▍     | 66/150 [02:39<01:54,  1.36s/it]

Normal_Videos_928_x264


 45%|████▍     | 67/150 [02:40<01:37,  1.18s/it]

Normal_Videos_892_x264


 45%|████▌     | 68/150 [02:41<01:43,  1.26s/it]

Normal_Videos_018_x264


 46%|████▌     | 69/150 [02:42<01:34,  1.17s/it]

Normal_Videos_168_x264


 47%|████▋     | 70/150 [02:44<01:39,  1.25s/it]

Normal_Videos_926_x264


 47%|████▋     | 71/150 [02:45<01:44,  1.32s/it]

Normal_Videos_417_x264


 48%|████▊     | 72/150 [02:46<01:32,  1.19s/it]

Normal_Videos_873_x264


 49%|████▊     | 73/150 [02:47<01:38,  1.27s/it]

Normal_Videos_930_x264


 49%|████▉     | 74/150 [02:50<02:10,  1.72s/it]

Normal_Videos_041_x264


 50%|█████     | 75/150 [02:51<01:53,  1.51s/it]

Normal_Videos_899_x264


 51%|█████     | 76/150 [02:52<01:43,  1.40s/it]

Normal_Videos_828_x264


 51%|█████▏    | 77/150 [02:53<01:28,  1.21s/it]

Normal_Videos_317_x264


 52%|█████▏    | 78/150 [02:54<01:18,  1.09s/it]

Normal_Videos_923_x264


 53%|█████▎    | 79/150 [03:11<06:54,  5.83s/it]

Normal_Videos_702_x264


 53%|█████▎    | 80/150 [03:14<05:45,  4.94s/it]

Normal_Videos_310_x264


 54%|█████▍    | 81/150 [03:16<04:41,  4.08s/it]

Normal_Videos_710_x264


 55%|█████▍    | 82/150 [03:17<03:44,  3.30s/it]

Normal_Videos_704_x264


 55%|█████▌    | 83/150 [03:19<03:02,  2.73s/it]

Normal_Videos_246_x264


 56%|█████▌    | 84/150 [03:23<03:28,  3.16s/it]

Normal_Videos_606_x264


 57%|█████▋    | 85/150 [03:24<02:43,  2.51s/it]

Normal_Videos_871_x264


 57%|█████▋    | 86/150 [03:27<03:02,  2.85s/it]

Normal_Videos_866_x264


 58%|█████▊    | 87/150 [03:28<02:24,  2.29s/it]

Normal_Videos_189_x264


 59%|█████▊    | 88/150 [03:29<01:50,  1.79s/it]

Normal_Videos_067_x264


 59%|█████▉    | 89/150 [03:30<01:32,  1.52s/it]

Normal_Videos_722_x264


 60%|██████    | 90/150 [03:37<03:15,  3.25s/it]

Normal_Videos_896_x264


 61%|██████    | 91/150 [03:39<02:47,  2.85s/it]

Normal_Videos_251_x264


 61%|██████▏   | 92/150 [03:39<02:01,  2.09s/it]

Normal_Videos_003_x264


 62%|██████▏   | 93/150 [03:42<02:03,  2.16s/it]

Normal_Videos_224_x264


 63%|██████▎   | 94/150 [03:48<03:02,  3.25s/it]

Normal_Videos_883_x264


 63%|██████▎   | 95/150 [03:48<02:09,  2.36s/it]

Normal_Videos_217_x264


 64%|██████▍   | 96/150 [03:49<01:53,  2.10s/it]

Normal_Videos_010_x264


 65%|██████▍   | 97/150 [03:50<01:31,  1.73s/it]

Normal_Videos_869_x264


 65%|██████▌   | 98/150 [03:52<01:33,  1.80s/it]

Normal_Videos_070_x264


 66%|██████▌   | 99/150 [03:53<01:16,  1.51s/it]

Normal_Videos_905_x264


 67%|██████▋   | 100/150 [03:54<01:07,  1.35s/it]

Normal_Videos_927_x264


 67%|██████▋   | 101/150 [03:55<01:05,  1.34s/it]

Normal_Videos_888_x264


 68%|██████▊   | 102/150 [03:56<00:52,  1.08s/it]

Normal_Videos_051_x264


 69%|██████▊   | 103/150 [03:58<01:02,  1.34s/it]

Normal_Videos_360_x264


 69%|██████▉   | 104/150 [03:59<00:54,  1.18s/it]

Normal_Videos_894_x264


 70%|███████   | 105/150 [04:01<01:05,  1.46s/it]

Normal_Videos_059_x264


 71%|███████▏  | 107/150 [04:02<00:46,  1.09s/it]

Normal_Videos_881_x264
Normal_Videos_870_x264


 72%|███████▏  | 108/150 [04:03<00:38,  1.10it/s]

Normal_Videos_452_x264


 73%|███████▎  | 109/150 [04:03<00:30,  1.34it/s]

Normal_Videos_478_x264


 73%|███████▎  | 110/150 [04:07<01:06,  1.65s/it]

Normal_Videos_902_x264


 74%|███████▍  | 111/150 [04:08<00:58,  1.50s/it]

Normal_Videos_879_x264


 75%|███████▍  | 112/150 [04:09<00:50,  1.33s/it]

Normal_Videos_034_x264


 75%|███████▌  | 113/150 [04:10<00:46,  1.26s/it]

Normal_Videos_913_x264


 76%|███████▌  | 114/150 [04:11<00:37,  1.03s/it]

Normal_Videos_943_x264


 77%|███████▋  | 115/150 [04:12<00:34,  1.02it/s]

Normal_Videos_944_x264


 77%|███████▋  | 116/150 [04:18<01:26,  2.53s/it]

Normal_Videos_641_x264


 78%|███████▊  | 117/150 [04:21<01:28,  2.68s/it]

Normal_Videos_876_x264


 79%|███████▊  | 118/150 [04:21<01:03,  1.97s/it]

Normal_Videos_929_x264


 79%|███████▉  | 119/150 [04:22<00:49,  1.61s/it]

Normal_Videos_783_x264


 80%|████████  | 120/150 [04:30<01:46,  3.55s/it]

Normal_Videos_906_x264


 81%|████████  | 121/150 [04:30<01:17,  2.66s/it]

Normal_Videos_937_x264


 81%|████████▏ | 122/150 [04:31<00:59,  2.14s/it]

Normal_Videos_621_x264


 82%|████████▏ | 123/150 [04:35<01:13,  2.71s/it]

Normal_Videos_312_x264


 83%|████████▎ | 124/150 [04:36<00:57,  2.21s/it]

Normal_Videos_686_x264


 83%|████████▎ | 125/150 [04:38<00:53,  2.14s/it]

Normal_Videos_175_x264


 84%|████████▍ | 126/150 [04:46<01:31,  3.80s/it]

Normal_Videos_909_x264


 85%|████████▍ | 127/150 [04:47<01:06,  2.88s/it]

Normal_Videos_603_x264


 85%|████████▌ | 128/150 [04:50<01:02,  2.83s/it]

Normal_Videos_048_x264


 86%|████████▌ | 129/150 [04:51<00:50,  2.39s/it]

Normal_Videos_798_x264


 87%|████████▋ | 130/150 [04:56<01:05,  3.27s/it]

Normal_Videos_597_x264


 87%|████████▋ | 131/150 [04:59<00:58,  3.07s/it]

Normal_Videos_940_x264


 88%|████████▊ | 132/150 [05:39<04:15, 14.17s/it]

Normal_Videos_210_x264


 89%|████████▊ | 133/150 [05:44<03:13, 11.35s/it]

Normal_Videos_874_x264


 89%|████████▉ | 134/150 [05:47<02:24,  9.03s/it]

Normal_Videos_924_x264


 90%|█████████ | 135/150 [07:31<09:19, 37.28s/it]

Normal_Videos_150_x264


 91%|█████████ | 136/150 [07:31<06:08, 26.34s/it]

Normal_Videos_933_x264


 91%|█████████▏| 137/150 [07:33<04:05, 18.88s/it]

Normal_Videos_890_x264


 92%|█████████▏| 138/150 [07:36<02:49, 14.09s/it]

Normal_Videos_063_x264


 93%|█████████▎| 139/150 [07:36<01:49,  9.96s/it]

Normal_Videos_877_x264


 93%|█████████▎| 140/150 [07:44<01:34,  9.41s/it]

Normal_Videos_656_x264


 95%|█████████▍| 142/150 [07:46<00:39,  4.99s/it]

Normal_Videos_345_x264
Normal_Videos_027_x264


 95%|█████████▌| 143/150 [07:50<00:33,  4.74s/it]

Normal_Videos_248_x264


 96%|█████████▌| 144/150 [07:51<00:21,  3.61s/it]

Normal_Videos_880_x264


 97%|█████████▋| 145/150 [08:06<00:34,  6.91s/it]

Normal_Videos_915_x264


 97%|█████████▋| 146/150 [08:07<00:20,  5.15s/it]

Normal_Videos_932_x264


 98%|█████████▊| 147/150 [08:08<00:12,  4.05s/it]

Normal_Videos_891_x264


 99%|█████████▊| 148/150 [08:10<00:06,  3.29s/it]

Normal_Videos_885_x264


 99%|█████████▉| 149/150 [08:10<00:02,  2.42s/it]

Normal_Videos_935_x264


  0%|          | 0/50 [00:00<?, ?it/s]

Fighting025_x264


  2%|▏         | 1/50 [00:03<03:12,  3.92s/it]

Fighting003_x264


  4%|▍         | 2/50 [00:06<02:29,  3.11s/it]

Fighting012_x264


  6%|▌         | 3/50 [00:07<01:39,  2.12s/it]

Fighting042_x264


  8%|▊         | 4/50 [00:09<01:33,  2.03s/it]

Fighting050_x264


 10%|█         | 5/50 [00:32<07:17,  9.72s/it]

Fighting027_x264


 12%|█▏        | 6/50 [00:35<05:30,  7.51s/it]

Fighting005_x264


 14%|█▍        | 7/50 [00:37<03:58,  5.54s/it]

Fighting035_x264


 16%|█▌        | 8/50 [00:38<02:49,  4.03s/it]

Fighting017_x264


 18%|█▊        | 9/50 [00:39<02:15,  3.31s/it]

Fighting020_x264


 20%|██        | 10/50 [00:41<01:53,  2.85s/it]

Fighting039_x264


 22%|██▏       | 11/50 [00:48<02:37,  4.04s/it]

Fighting030_x264


 24%|██▍       | 12/50 [00:50<02:08,  3.37s/it]

Fighting037_x264


 26%|██▌       | 13/50 [00:53<02:07,  3.45s/it]

Fighting051_x264


 28%|██▊       | 14/50 [00:59<02:26,  4.07s/it]

Fighting034_x264


 30%|███       | 15/50 [01:04<02:35,  4.44s/it]

Fighting026_x264


 32%|███▏      | 16/50 [01:07<02:15,  3.98s/it]

Fighting013_x264


 34%|███▍      | 17/50 [01:08<01:43,  3.14s/it]

Fighting028_x264


 36%|███▌      | 18/50 [01:10<01:29,  2.79s/it]

Fighting049_x264


 38%|███▊      | 19/50 [01:12<01:12,  2.34s/it]

Fighting006_x264


 40%|████      | 20/50 [01:12<00:56,  1.88s/it]

Fighting011_x264


 42%|████▏     | 21/50 [01:19<01:37,  3.38s/it]

Fighting024_x264


 44%|████▍     | 22/50 [01:25<01:52,  4.00s/it]

Fighting043_x264


 46%|████▌     | 23/50 [01:27<01:36,  3.58s/it]

Fighting015_x264


 48%|████▊     | 24/50 [01:36<02:13,  5.14s/it]

Fighting032_x264


 50%|█████     | 25/50 [01:38<01:41,  4.07s/it]

Fighting044_x264


 52%|█████▏    | 26/50 [01:45<01:58,  4.92s/it]

Fighting031_x264


 54%|█████▍    | 27/50 [01:52<02:13,  5.80s/it]

Fighting021_x264


 56%|█████▌    | 28/50 [01:55<01:49,  4.98s/it]

Fighting007_x264


 58%|█████▊    | 29/50 [01:59<01:33,  4.46s/it]

Fighting002_x264


 60%|██████    | 30/50 [02:01<01:15,  3.79s/it]

Fighting046_x264


 62%|██████▏   | 31/50 [02:03<01:04,  3.40s/it]

Fighting010_x264


 64%|██████▍   | 32/50 [02:08<01:09,  3.88s/it]

Fighting014_x264


 66%|██████▌   | 33/50 [02:12<01:06,  3.90s/it]

Fighting004_x264


 68%|██████▊   | 34/50 [02:27<01:53,  7.09s/it]

Fighting041_x264


 70%|███████   | 35/50 [02:54<03:16, 13.13s/it]

Fighting009_x264


 72%|███████▏  | 36/50 [02:56<02:15,  9.71s/it]

Fighting033_x264


 74%|███████▍  | 37/50 [02:57<01:32,  7.08s/it]

Fighting016_x264


 76%|███████▌  | 38/50 [02:59<01:08,  5.67s/it]

Fighting029_x264


 78%|███████▊  | 39/50 [03:03<00:56,  5.10s/it]

Fighting038_x264


 80%|████████  | 40/50 [03:04<00:37,  3.77s/it]

Fighting045_x264


 82%|████████▏ | 41/50 [03:09<00:37,  4.12s/it]

Fighting019_x264


 84%|████████▍ | 42/50 [03:11<00:29,  3.69s/it]

Fighting018_x264


 86%|████████▌ | 43/50 [03:12<00:20,  2.93s/it]

Fighting036_x264


 88%|████████▊ | 44/50 [03:16<00:17,  2.99s/it]

Fighting048_x264


 90%|█████████ | 45/50 [03:21<00:19,  3.85s/it]

Fighting022_x264


 92%|█████████▏| 46/50 [03:25<00:15,  3.90s/it]

Fighting023_x264


 94%|█████████▍| 47/50 [03:26<00:09,  3.02s/it]

Fighting047_x264


 96%|█████████▌| 48/50 [03:35<00:09,  4.64s/it]

Fighting008_x264


 98%|█████████▊| 49/50 [03:50<00:07,  7.70s/it]

Fighting040_x264


  0%|          | 0/50 [00:00<?, ?it/s]

Explosion016_x264


  2%|▏         | 1/50 [00:00<00:39,  1.25it/s]

Explosion018_x264


  4%|▍         | 2/50 [00:02<00:52,  1.09s/it]

Explosion044_x264


  6%|▌         | 3/50 [00:02<00:40,  1.15it/s]

Explosion037_x264


  8%|▊         | 4/50 [00:03<00:39,  1.16it/s]

Explosion011_x264


 10%|█         | 5/50 [00:04<00:46,  1.02s/it]

Explosion024_x264


 12%|█▏        | 6/50 [00:05<00:45,  1.03s/it]

Explosion035_x264


 14%|█▍        | 7/50 [00:08<01:03,  1.47s/it]

Explosion041_x264


 16%|█▌        | 8/50 [00:11<01:18,  1.88s/it]

Explosion012_x264


 18%|█▊        | 9/50 [00:11<01:00,  1.49s/it]

Explosion001_x264


 20%|██        | 10/50 [00:12<00:47,  1.19s/it]

Explosion027_x264


 22%|██▏       | 11/50 [00:12<00:39,  1.02s/it]

Explosion022_x264


 24%|██▍       | 12/50 [00:15<01:01,  1.62s/it]

Explosion034_x264


 26%|██▌       | 13/50 [00:16<00:45,  1.23s/it]

Explosion043_x264


 28%|██▊       | 14/50 [00:22<01:44,  2.91s/it]

Explosion003_x264


 30%|███       | 15/50 [00:23<01:16,  2.18s/it]

Explosion030_x264


 32%|███▏      | 16/50 [00:26<01:19,  2.35s/it]

Explosion004_x264


 34%|███▍      | 17/50 [00:28<01:17,  2.35s/it]

Explosion015_x264


 36%|███▌      | 18/50 [00:29<01:05,  2.04s/it]

Explosion045_x264


 38%|███▊      | 19/50 [00:30<00:53,  1.74s/it]

Explosion017_x264


 40%|████      | 20/50 [00:32<00:50,  1.69s/it]

Explosion009_x264


 42%|████▏     | 21/50 [00:33<00:42,  1.46s/it]

Explosion006_x264


 44%|████▍     | 22/50 [00:34<00:41,  1.49s/it]

Explosion038_x264


 46%|████▌     | 23/50 [00:37<00:47,  1.75s/it]

Explosion025_x264


 48%|████▊     | 24/50 [00:37<00:35,  1.35s/it]

Explosion008_x264


 50%|█████     | 25/50 [00:39<00:34,  1.37s/it]

Explosion002_x264


 52%|█████▏    | 26/50 [00:42<00:47,  1.98s/it]

Explosion047_x264


 54%|█████▍    | 27/50 [00:43<00:39,  1.73s/it]

Explosion042_x264


 56%|█████▌    | 28/50 [00:45<00:41,  1.90s/it]

Explosion040_x264


 58%|█████▊    | 29/50 [00:49<00:52,  2.51s/it]

Explosion046_x264


 60%|██████    | 30/50 [03:03<13:56, 41.81s/it]

Explosion051_x264


 62%|██████▏   | 31/50 [03:05<09:30, 30.04s/it]

Explosion052_x264


 64%|██████▍   | 32/50 [03:08<06:34, 21.92s/it]

Explosion020_x264


 66%|██████▌   | 33/50 [03:10<04:26, 15.66s/it]

Explosion023_x264


 68%|██████▊   | 34/50 [03:12<03:04, 11.56s/it]

Explosion021_x264


 70%|███████   | 35/50 [03:12<02:04,  8.30s/it]

Explosion019_x264


 72%|███████▏  | 36/50 [03:13<01:22,  5.91s/it]

Explosion028_x264


 74%|███████▍  | 37/50 [03:14<00:59,  4.58s/it]

Explosion026_x264


 76%|███████▌  | 38/50 [03:16<00:46,  3.88s/it]

Explosion013_x264


 78%|███████▊  | 39/50 [03:19<00:39,  3.57s/it]

Explosion007_x264


 80%|████████  | 40/50 [03:33<01:06,  6.65s/it]

Explosion036_x264


 82%|████████▏ | 41/50 [03:37<00:53,  5.96s/it]

Explosion010_x264


 84%|████████▍ | 42/50 [03:39<00:38,  4.79s/it]

Explosion032_x264


 86%|████████▌ | 43/50 [03:41<00:27,  3.87s/it]

Explosion014_x264


 88%|████████▊ | 44/50 [03:42<00:18,  3.07s/it]

Explosion033_x264


 90%|█████████ | 45/50 [03:46<00:15,  3.14s/it]

Explosion005_x264


 92%|█████████▏| 46/50 [03:47<00:10,  2.60s/it]

Explosion050_x264


 94%|█████████▍| 47/50 [03:48<00:06,  2.01s/it]

Explosion029_x264


 96%|█████████▌| 48/50 [03:52<00:05,  2.82s/it]

Explosion048_x264


 98%|█████████▊| 49/50 [03:53<00:02,  2.29s/it]

Explosion039_x264


  0%|          | 0/50 [00:00<?, ?it/s]

Arrest021_x264


  2%|▏         | 1/50 [00:07<06:22,  7.81s/it]

Arrest041_x264


  4%|▍         | 2/50 [00:11<04:12,  5.26s/it]

Arrest005_x264


  6%|▌         | 3/50 [00:14<03:20,  4.26s/it]

Arrest043_x264


  8%|▊         | 4/50 [00:23<04:43,  6.17s/it]

Arrest003_x264


 10%|█         | 5/50 [00:26<03:42,  4.94s/it]

Arrest028_x264


 12%|█▏        | 6/50 [00:28<02:53,  3.95s/it]

Arrest048_x264


 14%|█▍        | 7/50 [00:31<02:41,  3.76s/it]

Arrest020_x264


 16%|█▌        | 8/50 [00:34<02:20,  3.34s/it]

Arrest039_x264


 18%|█▊        | 9/50 [00:49<04:46,  7.00s/it]

Arrest018_x264


 20%|██        | 10/50 [00:51<03:37,  5.45s/it]

Arrest012_x264


 22%|██▏       | 11/50 [01:04<05:03,  7.77s/it]

Arrest040_x264


 24%|██▍       | 12/50 [01:05<03:42,  5.85s/it]

Arrest051_x264


 26%|██▌       | 13/50 [01:14<04:09,  6.74s/it]

Arrest002_x264


 28%|██▊       | 14/50 [01:16<03:14,  5.39s/it]

Arrest013_x264


 30%|███       | 15/50 [01:17<02:21,  4.04s/it]

Arrest024_x264


 32%|███▏      | 16/50 [01:20<02:06,  3.73s/it]

Arrest032_x264


 34%|███▍      | 17/50 [01:21<01:40,  3.04s/it]

Arrest044_x264


 36%|███▌      | 18/50 [01:24<01:34,  2.97s/it]

Arrest050_x264


 38%|███▊      | 19/50 [01:33<02:24,  4.66s/it]

Arrest019_x264


 40%|████      | 20/50 [01:34<01:48,  3.61s/it]

Arrest017_x264


 42%|████▏     | 21/50 [01:37<01:38,  3.39s/it]

Arrest023_x264


 44%|████▍     | 22/50 [01:38<01:18,  2.82s/it]

Arrest007_x264


 46%|████▌     | 23/50 [01:41<01:15,  2.79s/it]

Arrest037_x264


 48%|████▊     | 24/50 [01:43<01:04,  2.48s/it]

Arrest010_x264


 50%|█████     | 25/50 [01:45<01:00,  2.41s/it]

Arrest009_x264


 52%|█████▏    | 26/50 [01:47<00:50,  2.12s/it]

Arrest025_x264


 54%|█████▍    | 27/50 [01:47<00:40,  1.75s/it]

Arrest047_x264


 56%|█████▌    | 28/50 [02:46<06:54, 18.83s/it]

Arrest022_x264


 58%|█████▊    | 29/50 [02:48<04:46, 13.63s/it]

Arrest014_x264


 60%|██████    | 30/50 [02:49<03:18,  9.93s/it]

Arrest030_x264


 62%|██████▏   | 31/50 [02:56<02:52,  9.10s/it]

Arrest046_x264


 64%|██████▍   | 32/50 [03:03<02:29,  8.32s/it]

Arrest031_x264


 66%|██████▌   | 33/50 [03:04<01:46,  6.28s/it]

Arrest038_x264


 68%|██████▊   | 34/50 [03:06<01:19,  4.95s/it]

Arrest034_x264


 70%|███████   | 35/50 [03:16<01:35,  6.37s/it]

Arrest027_x264


 72%|███████▏  | 36/50 [03:17<01:08,  4.91s/it]

Arrest036_x264


 74%|███████▍  | 37/50 [03:20<00:57,  4.39s/it]

Arrest026_x264


 76%|███████▌  | 38/50 [03:26<00:56,  4.72s/it]

Arrest008_x264


 78%|███████▊  | 39/50 [03:31<00:54,  4.95s/it]

Arrest016_x264


 80%|████████  | 40/50 [03:40<01:01,  6.13s/it]

Arrest011_x264


 82%|████████▏ | 41/50 [03:42<00:43,  4.84s/it]

Arrest049_x264


 84%|████████▍ | 42/50 [03:52<00:50,  6.37s/it]

Arrest042_x264


 86%|████████▌ | 43/50 [03:53<00:33,  4.85s/it]

Arrest001_x264


 88%|████████▊ | 44/50 [03:55<00:23,  3.98s/it]

Arrest015_x264


 90%|█████████ | 45/50 [03:56<00:15,  3.03s/it]

Arrest035_x264


 92%|█████████▏| 46/50 [04:04<00:18,  4.57s/it]

Arrest006_x264


 94%|█████████▍| 47/50 [04:11<00:15,  5.15s/it]

Arrest004_x264


 96%|█████████▌| 48/50 [04:14<00:09,  4.57s/it]

Arrest029_x264


 98%|█████████▊| 49/50 [04:22<00:05,  5.50s/it]

Arrest033_x264


100%|██████████| 50/50 [04:29<00:00,  5.38s/it]


In [9]:
# 변환된 npy 파일 개수 확인
for i in dirlist[1:]:
    print(i.split('/')[2], len(os.listdir(i)))

Abuse 50
Assault 50
Arson 50
Burglary 100
Normal_Videos-Part-1 150
Fighting 50
Explosion 50
Arrest 50


## 1.2 Numpy Array Split with Uniform Sampling


In [4]:
import numpy as np
import os
from tqdm import tqdm

#### 시드 설정

In [82]:
np.random.seed(2045)    

### 1.2.1 Numpy Array Split

In [83]:
def NpArraySplit(path, train_rate, val_rate, test_rate): # Data split -> 파일명, 파일 경로
    # ex) NpArraySplit('./NP', 60,20,20):

    train_videos = []
    train_path= []
    val_videos = []
    val_path = []
    test_videos = []
    test_path = []

    for dirname, _, filenames in os.walk(path):
        n = len(filenames)
        
        for i,filename in enumerate(filenames):
            if(i < (n * train_rate / 100)): # train data : 60%
                train_videos.append(os.path.join(filename))
                train_path.append(os.path.join(dirname,filename))
            
            elif(i < (n * (train_rate + val_rate) / 100)): # valid data : 20%
                val_videos.append(os.path.join(filename))
                val_path.append(os.path.join(dirname,filename))

            else: # test data : 20%
                test_videos.append(os.path.join(filename))
                test_path.append(os.path.join(dirname,filename))

    return train_videos,train_path,val_videos,val_path,test_videos, test_path

In [84]:
train_videos, train_path, val_videos, val_path, test_videos, test_path = NpArraySplit('./NP', 60,20,20)

In [85]:
len(train_videos), len(val_videos), len(test_videos)

(330, 110, 110)

In [86]:
import pandas as pd
import numpy as np
import re

pattern = re.compile('[a-zA-Z]*')

def mkDataFrame(videos, path):
    # creating a dataframe having video names
    train = pd.DataFrame()
    train['video_name'] = videos
    train['path'] = path

    # creating tags for training videos
    train_video_tag = []
    for i in range(train.shape[0]):
        train_video_tag.append(pattern.match(train['video_name'][i]).group())

    train['class'] = train_video_tag

    return train

In [87]:
train = mkDataFrame(train_videos, train_path)
val = mkDataFrame(val_videos, val_path)
test = mkDataFrame(test_videos, test_path)

# DataFrame Shuffling
train = train.iloc[np.random.permutation(train.index)].reset_index(drop=True)
val = val.iloc[np.random.permutation(val.index)].reset_index(drop=True)
test = test.iloc[np.random.permutation(test.index)].reset_index(drop=True)

In [88]:
train.head()

video_name  \
0  Normal_Videos_439_x264.npy   
1         Assault024_x264.npy   
2        Burglary084_x264.npy   
3          Arrest041_x264.npy   
4          Arrest022_x264.npy   

                                                path     class  
0  ./NP/Normal_Videos-Part-1/Normal_Videos_439_x2...    Normal  
1                   ./NP/Assault/Assault024_x264.npy   Assault  
2                 ./NP/Burglary/Burglary084_x264.npy  Burglary  
3                     ./NP/Arrest/Arrest041_x264.npy    Arrest  
4                     ./NP/Arrest/Arrest022_x264.npy    Arrest

In [89]:
print(train.shape)
print(val.shape)
print(test.shape)

(330, 3)
(110, 3)
(110, 3)


### 1.2.2 Class reduction 8 -> 4

4개 클래스로 통합(500)
- **Normal** : 150
- **Arson** (+ Explosion) : 50 + 50 = 100
- **Assault** (+ Abuse, Fighting) : 50 + 50 + 50 = 150
- **Burglary** : 100

train val test
- train : 300
- val : 100
- test : 100

In [90]:
classes = ['Normal', 'Arson', 'Assault', 'Burglary']

In [91]:
train['class'].value_counts()

Normal       90
Burglary     60
Assault      30
Explosion    30
Abuse        30
Fighting     30
Arrest       30
Arson        30
Name: class, dtype: int64

In [92]:
train.replace({'class':'Explosion'}, {'class':'Arson'}, inplace=True)
train.replace({'class':['Abuse', 'Fighting']}, {'class':'Assault'}, inplace=True)
train.drop(train[train['class'] == 'Arrest'].index, axis=0, inplace=True)
train.reset_index(drop=True, inplace=True)

train['class'].value_counts()

Normal      90
Assault     90
Burglary    60
Arson       60
Name: class, dtype: int64

In [93]:
val.replace({'class':'Explosion'}, {'class':'Arson'}, inplace=True)
val.replace({'class':['Abuse', 'Fighting']}, {'class':'Assault'}, inplace=True)
val.drop(val[val['class'] == 'Arrest'].index, axis=0, inplace=True)
val.reset_index(drop=True, inplace=True)

val['class'].value_counts()

Normal      30
Assault     30
Burglary    20
Arson       20
Name: class, dtype: int64

In [94]:
test.replace({'class':'Explosion'}, {'class':'Arson'}, inplace=True)
test.replace({'class':['Abuse', 'Fighting']}, {'class':'Assault'}, inplace=True)
test.drop(test[test['class'] == 'Arrest'].index, axis=0, inplace=True)
test.reset_index(drop=True, inplace=True)

test['class'].value_counts()

Normal      30
Assault     30
Burglary    20
Arson       20
Name: class, dtype: int64

### 1.2.3 Numpy Array Load -> Uniform Sampling

In [95]:
 def uniform_sampling(video, target_frames):
        # get total frames of input video and calculate sampling interval 
        len_frames = int(len(video))
        interval = int(np.ceil(len_frames/target_frames))

        # init empty list for sampled video 
        sampled_video = []
        for i in range(0, len_frames, interval):
            sampled_video.append(video[i])     

        # calculate numer of padded frames and fix it 
        num_pad = target_frames - len(sampled_video)
        if num_pad > 0:
            padding = [video[i] for i in range(-num_pad, 0)]
            sampled_video += padding     

        # get sampled video
        return np.array(sampled_video, dtype=np.float32)

In [96]:
def Load_NP(data, target_frames):
    # creating an empty list
    data_image = []
    data_label = []

    # for loop to read and store frames
    for i in tqdm(range(data.shape[0])):
        video = np.load(data['path'][i])
        video = uniform_sampling(video, target_frames)

        # appending the image to the train_image list
        data_image.append(video)
        data_label.append(data['class'][i])

    return np.array(data_image), np.array(data_label)

In [97]:
def mk_train_val_test(train, val, test, target_frames):
    X_train, y_train = Load_NP(train, target_frames)
    X_val, y_val = Load_NP(val, target_frames)
    X_test, y_test = Load_NP(test, target_frames)
    
    print(X_train.shape, y_train.shape)
    print(X_val.shape, y_val.shape)
    print(X_test.shape, y_test.shape)
    
    return X_train, y_train, X_val, y_val, X_test, y_test

In [98]:
X_train, y_train, X_val, y_val, X_test, y_test = mk_train_val_test(train, val, test, 64)

100%|██████████| 100/100 [03:31<00:00,  2.11s/it]


(300, 64, 128, 128, 3) (300,)
(100, 64, 128, 128, 3) (100,)
(100, 64, 128, 128, 3) (100,)


### 1.2.4 y -> One-Hot Encoding

In [99]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

y_train_LE = encoder.fit_transform(y_train)
y_val_LE = encoder.fit_transform(y_val)
y_test_LE = encoder.fit_transform(y_test)

In [100]:
from keras.utils import to_categorical

y_train_OHE = to_categorical(y_train_LE)
y_val_OHE = to_categorical(y_val_LE)
y_test_OHE = to_categorical(y_test_LE)

In [101]:
y_train_OHE

array([[0., 0., 0., 1.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       ...,
       [0., 1., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 0., 1.]], dtype=float32)

### 1.2.5 Train, Val, Test -> Numpy Array Save

In [102]:
save_dir = 'Data_class_integrate_4'

if not os.path.exists(save_dir):
    os.makedirs(save_dir)
    
    np.save(save_dir + '/X_train.npy', X_train)
    np.save(save_dir + '/y_train.npy', y_train_OHE)
    np.save(save_dir + '/X_val.npy', X_val)
    np.save(save_dir + '/y_val.npy', y_val_OHE)
    np.save(save_dir + '/X_test.npy', X_test)
    np.save(save_dir + '/y_test.npy', y_test_OHE)
    
else:
    print('Already done')